### PR TITLE
Featre: reporting bc

### PR DESCRIPTION
**Pull request by pedrosousabarreto**
Initial version of the reporting BC, there some TODOs in the docs and a missing link to the not yet existing operation BC doc section.

### CODE REVIEW

File: docs/refarch/boundedContexts/reporting/index.md [hunk 1]
```diff
@@ -1,21 +1,66 @@
...
+
+This Reporting Data Store must be write-only from the switch's perspective and read-only by external components.
+
+The data models on the Reporting Data Store can be different from the internal operational data models that the switch uses; where relevant, for performance or other reasons, multiple models of the same data can be made available in the Reporting Data Store - akin to multiple projections or views.
+A Client BC Reporting Component provided by the switch, will be translating internal events and internal data models to the external data store models - This component can be replaced or, there can even exist many of these for a single Client BC.
+Such component must exists in the Reporting BC for any Client BC's who's data is made available in the Reporting Data Store. 
+Any direct data sending, or fetching, from the source Client BC, or its internal data stores, to the Reporting Data Store, constitutes a violation of the decoupling principle and will negatively affect the maintainability of the system by virtue of its tight coupling.
```
Comment (mdebarros) [inline docs/refarch/boundedContexts/reporting/index.md:10]: Can we clarify or expand on `Any direct data sending` please.

In the previous sentence, it indicates that we can do this with a Reporting component specific to each BC to send the data.

Perhaps by this, you mean that no Client BC should "write" directly to the Reporting BC's Data store?

**Review by mdebarros — APPROVED** [review]
+1

**Comment by millerabel** [comment]
Can we also clarify whether we mean Write Only or Append Only? Appending adds new data to the record; Writing implies overwriting (updating) the record of past reported data. 

The permanent record of what occurred (a journal) is a sequence of choices made in reaction to events. This can be implemented safely using Append Only interfaces, but becomes highly suspicious if writing is permitted.

— Miller

On Sep 27, 2021, at 6:13 AM, Miguel de Barros ***@***.***> wrote:

﻿
@mdebarros commented on this pull request.

In docs/refarch/boundedContexts/reporting/index.md:

>  
-{overview}
+The Reporting strategy for this Reference Architecture, is to describe the generic mechanisms by which client BCs' data can be persisted, and kept up to date, in a Reporting Data Store, so users and systems, can later consume this data directly from the Reporting Data Store, or through any Reporting and/or Dashboarding tools connected to the Reporting Data Store.  
+
+This Reporting Data Store must be write-only from the switch's perspective and read-only by external components.
+
+The data models on the Reporting Data Store can be different from the internal operational data models that the switch uses; where relevant, for performance or other reasons, multiple models of the same data can be made available in the Reporting Data Store - akin to multiple projections or views.
+A Client BC Reporting Component provided by the switch, will be translating internal events and internal data models to the external data store models - This component can be replaced or, there can even exist many of these for a single Client BC.
+Such component must exists in the Reporting BC for any Client BC's who's data is made available in the Reporting Data Store. 
+Any direct data sending, or fetching, from the source Client BC, or its internal data stores, to the Reporting Data Store, constitutes a violation of the decoupling principle and will negatively affect the maintainability of the system by virtue of its tight coupling.
Can we clarify or expand on Any direct data sending please.

In the previous sentence, it indicates that we can do this with a Reporting component specific to each BC to send the data.

Perhaps by this, you mean that no Client BC should "write" directly to the Reporting BC's Data store?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by pedrosousabarreto** [comment]
Good point Miller, in this case it is write only from the switch's perspective.

The logic is that these reporting components will be responsible for putting data in reporting data stores and keeping that up to date, so only those reporting components can have access to it, and only need write access. A possible exception might be if for the "keeping it up to date" part those reporting components need read access, but that's not likely.

The most important point is that we can never use the reporting data stores internally for reads.

This is still a draft, but check here: https://ref-arch-docs.moja-lab.live/refarch/boundedContexts/reporting/

**Pull request merged**